Support mixed case Kind

mariantalla opened the pull request:
This PR should address issues https://github.com/kubernetes-sigs/kubebuilder/issues/1019 and https://github.com/kubernetes-sigs/kubebuilder/issues/1091.

The implementation uses the validation logic that runs against `Kind` in
apimachinery; i.e. allows any `Kind` value that, when converted to
lowercase, is a valid DNS-1035 label.

Reusing that logic instead of reimplementing it helps with ensuring that
the two (i.e. `Kind` validation in core Kubernetes and `Kind` validation
in kubebuilder) are as similar as possible.

Would love feedback on:
* The choice to use [`k8s.io/apimachinery` logic](https://github.com/kubernetes/kubernetes/blob/b3b15a9e260ac65edfa14dd80add9046f3fa9b1f/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L549-L582) vs continuing to extend the validation logic [in kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/blob/702b272f9c73c723d0aec3569489ffcf79d6d1b9/pkg/scaffold/v1/resource/resource.go#L82).
* If the above does sound like a good idea, is it worth refactoring [`Validate()`](https://github.com/kubernetes-sigs/kubebuilder/blob/702b272f9c73c723d0aec3569489ffcf79d6d1b9/pkg/scaffold/v1/resource/resource.go#L56) more broadly, to use available logic in `k8s.io/apimachinery`?
* There were more changes than I expected in go.mod 😅 I'll take an extra look at that.

Remaining work in this PR:
(once the approach seems reasonable)
* Refactor tests to describe successful & failed validation more succinctly (at the moment there's duplication and missing test cases while the PR is in progress)
* Ensure any logging and error messages are descriptive of the validation that takes place.

mariantalla commented on go.mod:
```diff
@@ -4,18 +4,20 @@ go 1.12
+	google.golang.org/appengine v1.5.0 // indirect
+	gopkg.in/yaml.v2 v2.2.4
+	k8s.io/api v0.0.0-20190918195907-bd6ac527cfd2
+	k8s.io/apimachinery v0.0.0-20191004115801-a2eda9f80ab8
```
This does map to [`k8s.io/apimachinery`@`kubernetes-1.16.2`](https://github.com/kubernetes/apimachinery/releases/tag/kubernetes-1.16.2) (the latest stable at the moment), but I can't get go modules to save it like that. I could be missing something 🤔

mariantalla replied:
OK, from learning a bit more about go modules, that's because the tag names are not semver-style (i.e. `vx.y.z`). The commits correspond to the right tags though, so that should be fine.

mariantalla commented:
/hold
(while there's open TODO items)

mariantalla commented:
Any thoughts @mengqiy @DirectXMan12 @Liujingfang1 ?

/assign @mengqiy @DirectXMan12 @Liujingfang1

jdef commented on pkg/scaffold/v1/resource/resource.go:
```diff
@@ -79,7 +80,10 @@ func (r *Resource) Validate() error {
+	validationErrors := validation.IsDNS1035Label(strings.ToLower(r.Kind))
+
+	if len(validationErrors) != 0 {
 		return fmt.Errorf("kind must be camelcase (expected %s was %s)", flect.Pascalize(r.Kind), r.Kind)
```
there's still a call to `Pascalize` here; AFAICT this type of normalization should be eliminated.

jdef commented on pkg/scaffold/v1/resource/resource_test.go:
```diff
@@ -79,24 +79,19 @@ var _ = Describe("Resource", func() {
 		})
 
-		It("should fail if the Kind is not camel cased", func() {
+		It("should succeed if the Kind is camel cased", func() {
```
maybe include a negative test here too? something that fails dns1035 label validation, just so that we can see that validation error checks are working as expected.

mariantalla replied:
[This test](https://github.com/kubernetes-sigs/kubebuilder/pull/1133) on length and [this one](https://github.com/kubernetes-sigs/kubebuilder/blob/cc8d4ffed8f5d02be97c528998e581a7b8776e13/pkg/scaffold/resource/resource_test.go#L91) on allowed characters cover failed validations. Is there something missing? 🤔

camilamacedo86 commented on go.mod:
```diff
@@ -3,19 +3,17 @@ module sigs.k8s.io/kubebuilder
 go 1.12
 
 require (
```
Update the go.mo and go.sum shows not be part of the scope of this PR. 
I'd recommend you revert it and ensure that your branch is rebased with the master.

camilamacedo86 replied:
Hi @mariantalla,  

I still not getting why the modules need to be updated with these changes. So, I'd suggest you:

- Ensure that the branch is rebased with master
- Revert the changes of go.mod and go.sum
- Run `make generate`

mariantalla commented:
Thanks everyone for the comments. There's indeed a couple of follow-ups before this is on its way to merge (I've updated the PR description so it's clearer what work is left).

At the moment I was curious to hear how folks left towards using `k8s.io/apimachinery` code for validation (as opposed to continuing to extend the validation that kubebuilder is doing). Once there's consensus in that, error messages and test cases are pretty straightforward to change.

camilamacedo86 commented on pkg/scaffold/v1/resource/resource.go:
```diff
@@ -79,7 +80,10 @@ func (r *Resource) Validate() error {
 	}
-	if r.Kind != flect.Pascalize(r.Kind) {
+
+	validationErrors := validation.IsDNS1035Label(strings.ToLower(r.Kind))
```
See here that the import is not required. In the PR https://github.com/kubernetes-sigs/kubebuilder/pull/1048/files#diff-ccdb7c3b4336337dc2e653f1fbd88c7dR71 a similar validation from apimachinery was added in the project. There was decide do not do the import. So, I'd recommend you do the same.

mariantalla commented:
Following the advice in https://github.com/kubernetes-sigs/kubebuilder/pull/1048#discussion_r329712728 (thanks @camilamacedo86  for the pointer), we've removed dependencies on apimachinery.

Adirio commented:
The Travis error seems not to be related to this PR but to #1169

mariantalla commented:
oops, forgot to
/hold cancel

camilamacedo86 commented on pkg/scaffold/resource/resource.go:
```diff
@@ -96,9 +99,16 @@ const (
 	// dns1123SubdomainMaxLength is a subdomain's max length in DNS (RFC 1123)
 	dns1123SubdomainMaxLength int = 253
+
+	dns1035LabelFmt    string = "[a-z]([-a-z0-9]*[a-z0-9])?"
```
Shows that it is not the same validation for Kinds. Could we not have a kind == AppServer for example?

mariantalla replied:
Yep; the dns1035-like validation happens after the `Kind` [has been lowercased](https://github.com/kubernetes-sigs/kubebuilder/blob/5742ddd7a096570cd7fcc0881176d9dcb189d27b/pkg/scaffold/resource/resource.go#L83), so all of these combinations would be supported. apimachinery follows the same pattern, as far as I can tell (I'll find it and link here in a moment).

camilamacedo86 replied:
Hi @mariantalla, 

See [here](https://github.com/kubernetes/apimachinery/blob/6e68a40eebf94cb3e778d807bce5637d6aa079ea/pkg/apis/meta/v1/types_swagger_doc_generated.go#L57-L58) that  kubernetes API andd apimachinery will allow Kinds with upper case.

The validation that you got from there is not used to check Kinds.  You can use this one [here](https://github.com/kubernetes/apimachinery/blob/ced427e1ea5fdd2d307fa083f960961ceff71b65/pkg/util/validation/validation.go#L123 ) to verify the kinds.

DirectXMan12 replied:
I think @mariantalla means that apimachinery validates that the lower-case form of any kind is a valid dns-like label

DirectXMan12 replied:
(although we should have links in comments referencing that)

DirectXMan12 replied:
the relevant section: 

https://github.com/kubernetes/kubernetes/blob/b3b15a9e260ac65edfa14dd80add9046f3fa9b1f/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L549-L582

@mariantalla is correct

camilamacedo86 replied:
HI @mariantalla and @DirectXMan12,
 
Really tks for the explanations. I got that the lower case is just for the check and also, I tested locally the PR and all is working fine. It all shows fine, just the tests that were removed that IHMO should be back. We should still be ensuring that Crew and Guestbook and etc are accepted.

mariantalla replied:
Added the link in a few places in the conversation so that future-us can find it too :)

camilamacedo86 commented on pkg/scaffold/resource/resource_test.go:
```diff
@@ -71,24 +71,27 @@ var _ = Describe("Resource", func() {
-		It("should fail if the Kind is not pascal cased", func() {
-			// Base case
-			instance := &Resource{Group: "crew", Kind: "FirstMate", Version: "v1"}
+		It("should succeed if the Kind is valid, according to core Kubernetes", func() {
```
IHMO the existent KInds check should pass. 
Here for example : https://github.com/kubernetes-sigs/kubebuilder/issues/1019 is for we allow AWSCluster

mariantalla replied:
Agreed; this PR relaxes `Kind` validation (to bring it in line with the way it happens in core k8s). Previous passing tests continue to pass, and some previously failing should no longer fail (e.g. [this one](https://github.com/kubernetes-sigs/kubebuilder/blob/971cf03ea49aefc2380fdb6c5a91c55316f4617e/pkg/scaffold/resource/resource_test.go#L83-L91)). Or am I missing something? 🤔

camilamacedo86 replied:
The validation obtained is not the correct one for we check the kinds. If you use the correct one the only change here will be adding more one test to ensure that `AWSCluster` will pass. Also, note that is required to remove the pascalze as well since it is not something check-in k8s API too.

camilamacedo86 commented on pkg/scaffold/resource/resource.go:
```diff
@@ -79,8 +79,11 @@ func (r *Resource) Validate() error {
 			"version must match ^v\\d+(alpha\\d+|beta\\d+)?$ (was %s)", r.Version)
 	}
-	if r.Kind != flect.Pascalize(r.Kind) {
-		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(r.Kind), r.Kind)
```
My understand to solve the https://github.com/kubernetes-sigs/kubebuilder/issues/1091 is just replace the text in the msg `CamelCase` to `PascalCase` which shows done already.

mariantalla replied:
This PR will remove `Pascalize` completely though (in favor of the kube-like validation), and the new error messages describe what went wrong (hopefully! Open to feedback on that 😄 )

camilamacedo86 replied:
Solved can close this one. great work 👍

camilamacedo86 commented on pkg/scaffold/resource/resource.go:
```diff
@@ -79,8 +79,11 @@ func (r *Resource) Validate() error {
-	if r.Kind != flect.Pascalize(r.Kind) {
-		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(r.Kind), r.Kind)
+
+	validationErrors := isDNS1035Label(strings.ToLower(r.Kind))
```
my recommendation here to solve it is:
- Remove the Pascalize
- Check if the Kind matches with: 
`go
	ResourceKindRegexp = regexp.MustCompile("^[A-Z]{1}[a-zA-Z0-9]+$")
`

mariantalla commented:
Thanks for the comments @camilamacedo86 ! I've responded inline; but just to make sure I got your last suggestion right:

> Remove the Pascalize

I think this is happening [here](https://github.com/kubernetes-sigs/kubebuilder/pull/1133/files#diff-1264bafa536db5239815912ade8bc615L82-L83); where there other spots it needs to change?

> Use the func ... from apimachinery

Kind validation now uses IsDns1350Label, same as in core kubernetes

> Then, add a unit test

I think [this test](https://github.com/kubernetes-sigs/kubebuilder/blob/5742ddd7a096570cd7fcc0881176d9dcb189d27b/pkg/scaffold/resource/resource_test.go#L74-L77) covers that case, is there something missing?

jdef commented:
Maybe also include a test that rejects a Kind that's all lowercase? non-exported types don't seem very useful and the old Pascal validation used to guard against them.

camilamacedo86 commented:
HI @mariantalla, 

> validation now uses IsDns1350Label, same as in core Kubernetes  

I am able to apply KInds with upper cases in Kubernetes so it shows not be the validation used by it to check Kinds.

camilamacedo86 commented:
/assign @camilamacedo86

DirectXMan12 commented:
> Maybe also include a test that rejects a Kind that's all lowercase? non-exported types don't seem very useful and the old Pascal validation used to guard against them.

Yeah.  Technically this is allowed, practically we should prevent it.

DirectXMan12 commented:
otherwise looks good to me

DirectXMan12 commented on pkg/scaffold/resource/resource_test.go:
```diff
@@ -71,24 +71,27 @@ var _ = Describe("Resource", func() {
+			var kind string
 
-			instance = &Resource{Group: "crew", Kind: "firstMate", Version: "v1"}
-			Expect(instance.Validate()).NotTo(Succeed())
```
should bring these tests back though, so we have a full range

mariantalla replied:
Done ✅

DirectXMan12 replied:
table is fine :-)

camilamacedo86 commented on pkg/scaffold/resource/resource.go:
```diff
@@ -79,8 +79,11 @@ func (r *Resource) Validate() error {
+	validationErrors := isDNS1035Label(strings.ToLower(r.Kind))
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("Kind failed validation: %#v", validationErrors)
```
```suggestion
		return fmt.Errorf("Invalid Kind: %#v", validationErrors)
```

mariantalla replied:
Thanks, I added it.

mariantalla commented:
Pushed a couple of commits that I think cover any remaining points:
* added back previous test cases
* added a test and implementation to reject Kinds that have lowercase first characters

Adirio commented:
Travis is failing because the golden test files were not regenerated (and we changed year)

mariantalla commented:
> Travis is failing because the golden test files were not regenerated (and we changed year)

Oh yes... Thanks for the pointer!

Adirio commented:
#1282 should fix it once it gets merged

Adirio commented on pkg/scaffold/resource/resource.go:
```diff
@@ -73,9 +73,17 @@ func (r *Resource) Validate() error {
-	if r.Kind != flect.Pascalize(r.Kind) {
-		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(r.Kind), r.Kind)
+
+	validationErrors := []string{}
```
Just a minor comment? Could this be a `[]error{}`?

DirectXMan12 replied:
agreed with @Adirio  here -- make the the corresponding function return `[]error`

mariantalla replied:
Sure 👍 However, a lot of other functions in that file use the same pattern (collect error messages in a  `[]string`, return an error in the end with all the error messages)

e.g. (from current master): https://github.com/kubernetes-sigs/kubebuilder/blob/cf6cec42bdf594eac67cfb8f13eee6f57ba5c47b/pkg/scaffold/resource/resource.go#L126-L135

It traces back to `func maxLenError(length int) string {` returning a string, so it would be a broader change. It's pretty straightforward, just checking that it would be fine to include in this PR.

mariantalla replied:
(treats me right for saying `It's pretty straightforward` :) ) This change would also require refactoring test expectations slightly. 

e.g.  `Expect(err).To(MatchError(ContainSubstring("a DNS-1035 label must consist of lower case alphanumeric characters")))` 
would fail if err now holds a slice of `error`s rather than `string`s. Again, happy to go ahead with it, just making sure it would make sense.

DirectXMan12 replied:
oh, if the rest of the file does something similar, it's fine to leave for now :-).  Up to you.  Let me know what you decide.

mariantalla replied:
Left it as it was for this PR in the end. I will spend a bit of time right after to experiment, and will open a small cleanup PR if I figure out a way to work with `error`s in `resource.go` while keeping the test cases readable enough.

mariantalla commented on pkg/scaffold/resource/resource.go:
```diff
@@ -74,7 +74,13 @@ func (r *Resource) Validate() error {
-	validationErrors := isDNS1035Label(strings.ToLower(r.Kind))
+	validationErrors := []string{}
+
+	if string(r.Kind[0]) == strings.ToLower(string(r.Kind[0])) {
```
Is this line saying clearly enough "if Kind's first letter is lowercase"?

Kinda wanted to extract in its own function, but didn't think the benefit was worth the indirection 🤔 

@camilamacedo86 @Adirio @DirectXMan12

Adirio replied:
I would say it is pretty clear but a comment before it saying that we check that the first letter is lower case would not hurt anyone.

mariantalla replied:
Fixed, thanks! Resolving this one.

Adirio commented:
I'm expecting Travis to fail (do to an unrelated bug) but should be fixed after #1288 gets merged. LGTM otherwise.

mariantalla commented:
👋 Is there anything else missing here?

Adirio commented:
I think you may want to rebase to solve the Travis issue and benefit from all the added lint checks that were added in the meantime.

DirectXMan12 commented:
whoops, lost track of this.  Can merge after rebase.

/assign @mengqiy 
please do the rebase and we'll get this merge.

mengqiy commented:
@mariantalla Sorry for the delay on this PR.
Can you please rebase, so we can get this merged.
Some files have been refactored recently by @Adirio, it may require a little more work to rebase.

Adirio commented:
With the recent scaffolding it may be easier to branch from master and copy-paste your checks. `pkg/model/resource/options.go` should be where the check for Kind is placed, in the `Validate` method. `pkg/model/resource/options_test.go` should be where the additional tests go. Want some help?

mariantalla commented:
Thanks! I'll make a start at rebasing, will poke you if it all goes wrong @Adirio :)

mariantalla commented:
Did a `go mod tidy` too, which brought in the changes to `go.mod` and `go.sum`; let me know if we don't want these in.

Adirio commented:
> Did a `go mod tidy` too, which brought in the changes to `go.mod` and `go.sum`; let me know if we don't want these in.

We should probably add a test that checks that go mod tidy was done to Travis so that we do not forget about it.

Adirio commented on go.mod:
```diff
@@ -4,17 +4,19 @@ go 1.13
 	golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac
-	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
+	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
+	gopkg.in/yaml.v2 v2.2.4 // indirect
```
I would say that updating versions should not be performed in this PR unless you need one of them. Always updating to the last version makes tagging a version useless as next PR may change the version and it may break the code.

mariantalla replied:
Sounds good-checked that out from `master`.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
+			},
+			Entry("camelcase", "FirstMate"),
+			Entry("more than one caps at the start", "FIRSTMate"),
+		)
```
IMHO, I would not use `DescribeTable` here as all the rest of the tests are using raw `Describe`, `Context` and `It`. The added benefit of using `DescribeTable` also seems pretty low, as it is a 2LoC function and 2 entries. So I would just stick to what it is being used already. But thats just an opinion.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
+
+			for i := 0; i < 65; i++ {
+				kind += "a"
+			}
```
`kind := strings.Repeat("a", 64)`

mariantalla replied:
Thanks, changed it 👌

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
-			Expect(options.Validate()).NotTo(Succeed())
-			Expect(options.Validate().Error()).To(ContainSubstring(
-				`kind must be PascalCase (expected FirstMate was firstMate)`))
+		It("should fail if Kind is invalid, according to core Kubernetes", func() {
```
```suggestion
		It("should fail if Kind has whitespaces", func() {
```

mariantalla replied:
Good point; I've added a small collection of `Kind`s that would be considered invalid instead-of which whitespace is one. My reasoning here is that this test also serves as documentation for future us :) (or other humans that work on this code) on what logic we decided more broadly.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
+
+			options := &Options{Group: "crew", Kind: kind, Version: "v1"}
+			err := options.Validate()
+			Expect(err).To(MatchError(ContainSubstring("must be no more than 63 characters")))
```
I think you should add `Expect(err).To(HaveOccurred())` above this line to avoid panics in case the error is nil.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
+		It("should fail if Kind is invalid, according to core Kubernetes", func() {
+			options := &Options{Group: "crew", Kind: "Something withSpaces", Version: "v1"}
+			err := options.Validate()
+			Expect(err).To(MatchError(ContainSubstring("a DNS-1035 label must consist of lower case alphanumeric characters")))
```
I think you should add `Expect(err).To(HaveOccurred())` above this line to avoid panics in case the error is nil.

mariantalla replied:
`MatchError(...)` can handle `nil` errors gracefully from what I remember, so `HaveOccured()` is likely redundant 🤔 I'll double check

mariantalla replied:
Yep indeed (for cross reference: [gomega implementation](https://github.com/onsi/gomega/blob/6be6c439588487cd908a3a700795660c2a16dfec/matchers/match_error_matcher.go#L16-L18))

No super strong opinions though, happy to add it if there's significant value in adding the error expectation.

Adirio replied:
Didn't know about this helper until today so nice to learn something new (I should probably apply it in the `pkg/scaffold/internal/machienry` package too). Probably the error messages would be a bit different but they should both clear enough.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +74,37 @@ var _ = Describe("Resource Options", func() {
+		It("should fail if Kind starts with a lowercase character", func() {
+			options := &Options{Group: "crew", Kind: "lOWERCASESTART", Version: "v1"}
+			err := options.Validate()
+			Expect(err).To(MatchError(ContainSubstring("Kind must start with an uppercase character")))
```
I think you should add `Expect(err).To(HaveOccurred())` above this line to avoid panics in case the error is nil.

Adirio commented on pkg/model/resource/options_test.go:
```diff
@@ -73,24 +76,37 @@ var _ = Describe("Resource Options", func() {
+					ContainSubstring("a DNS-1035 label must consist of lower case alphanumeric characters")))
+			},
+			Entry("should fail validation if Kind contains whitespaces", "Something withSpaces"),
+			Entry("should fail validation if Kind ends in -", "Hello-"),
```
What about an extra `Entry("should fail validation if Kind starts with a number", "0Hello")`

mariantalla replied:
Good point, added one. Thanks!

Adirio commented:
Otherwise, LGTM.

mariantalla commented:
/retest
